src/js/common/parser.js

```diff
@@ -0,0 +1,223 @@
+    `["root", "nested", "value", "description"]`.
+
+ This function assumes it is working with an object that directly represents a JSON Schema, and that child properties
+ will be represented within the `properties` element of each level.  The path `.root.nested.value.description` implies
+ a structure like the following:
+
+    {
+      root: {
+        properties: {
+          nested: {
+            properties: {
+              value: {
+                description: "Text description"
+              }
+            }
+          }
+        }
+      }
+    }
+
+  At all but the last level, the segment (`nested`, for example`) is looked for first as a property, then as a root
+  element.  On the last level, the segment (`description`, for example) is looked for first as a root element, and then
+  as a property. This is done to avoid unexpected behavior in dealing with schema properties that match JSON Schema
+  keywords, such as `required`, `title`, or `description`.
+
+ */
+
+gpii.schema.parser.lookupField = function (that, schemaKey, schemaFieldPath) {
+    var pathSegments = Array.isArray(schemaFieldPath) ? schemaFieldPath : gpii.schema.validator.extractPathSegments(schemaFieldPath);
+
+    var dereferencedSchema = that.dereferencedSchemas[schemaKey];
+    if (dereferencedSchema) {
+        var currentLevel = dereferencedSchema;
+        for (var a = 0; a < pathSegments.length; a++) {
+            var segment = pathSegments[a];
+            var nextLevel = currentLevel.properties && currentLevel.properties[segment] ? currentLevel.properties[segment] : currentLevel[segment];
+            if (a === pathSegments.length - 1) {
+                nextLevel = currentLevel[segment] ? currentLevel[segment] : nextLevel;
+            }
+            if (nextLevel) {
+                currentLevel = nextLevel;
+            }
+            else {
+                return false;
+            }
+        }
+
+        return currentLevel;
+    }
+
+    // If we can't evolve the output, return `false` so that we can use the existing error.
+    return false;
+};
+
+/*
+
+  Convenience function to lookup the most commonly used field (the description).
+
+ */
+gpii.schema.parser.lookupDescription = function (that, schemaKey, schemaFieldPath) {
+    var pathPlusDescription = Array.isArray(schemaFieldPath) ? schemaFieldPath.concat("description") : schemaFieldPath + ".description";
+    return gpii.schema.parser.lookupField(that, schemaKey, pathPlusDescription);
+};
+
+
+
+/*
+
+  A listener to cache schemas as they are added.
+
+ */
+gpii.schema.parser.updateSchemas = function (that) {
+    var promises = [];
+    fluid.each(that.model.schemas, function (schemaContent, schemaKey) {
+        if (!that.dereferencedSchemas[schemaKey]) {
+            promises.push(gpii.schema.parser.dereference(that, schemaKey));
+        }
+    });
+
+    if (promises.length > 0) {
+        fluid.promise.sequence(promises).then(function () {
```

Comment:
https://github.com/the-t-in-rtf/gpii-json-schema/commit/d9e49e2c23f72c8c461dee9e064ab26a1d0ce84a